sysdep/unix/krt.c

```diff
@@ -657,7 +657,7 @@ krt_same_dest(rte *k, rte *e)
   switch (ka->dest)
     {
     case RTD_ROUTER:
-      return ipa_equal(ka->gw, ea->gw);
+      return ipa_equal(ka->gw, ea->gw) && ka->iface->index == ea->iface->index;
```

Comment:
This looks in the right ballpark to me.  Just some detail to consider...

Please compare with https://github.com/projectcalico/bird/commit/221a5d88ed7ebf0d28eb9e2b507a348b7504087f (PR https://github.com/projectcalico/bird/pull/38), which was a quite similar change.

- IIUC, that earlier change was for the case where a route is still *intended* to be a tunneled route.  Your change now is for the case where a previously tunneled route now needs reprogramming as a non-tunneled route.

- It makes sense that an outgoing interface (OIF) comparison is needed in both cases.  (Sorry for missing that in my earlier change.)

- The earlier change inlined the krt_same_dest condition (at the same time as adding OIF comparison), as @Dirbaio and I thought that was clearer.  I think it would be a good idea to do that here too.  (And then krt_same_dest can go, if it's not used anywhere else.)

- Detailed difference though: my earlier change got the wanted OIF by lookup of the krt_tunnel attribute, and compared by name; whereas this change implies that the wanted OIF is already available in new->attrs->iface, and compares by index.

I'd prefer if we could understand and square away that last difference.  I.e., I think, if the wanted OIF is already in new->attrs->iface for the tunneled case too, let's use that and compare by index in both cases, and lose the krt_tunnel lookup.

WDYT?

Reply:
@neiljerram :  It wasn't clear to me from a code read that `new` would include the correct interface in the tunnel scenario.  From what I can see the tunnel interface is "jimmied in" last minute when the route is programmed in `nl_send_route` in `netlink.c` - but perhaps that is merely constructing the correct structures at that point.